nm ipv4 ipv6: Fix DHCPv4 with static IPv6

cathay4t opened the pull request:
When DHCPv4 enabled with static IPv6, the verification stage will fail
due to the missing of static IPv6 configurations.
This is caused by we stop waiting once reach IP_CONFIG state where
static IP is not applied yet.
To fix this problem, nmstate should wait on fully activated when static
IP enabled.

ffmancera commented on libnmstate/nm/active_connection.py:
```diff
@@ -205,3 +197,23 @@ def _is_device_master_type(nmdev):
+        return False
+    if ipv6.is_static(nm_ac) and not ac_state_flags & nm_state_flags.IP6_READY:
+        return False
+    return ipv4.is_dynamic(nm_ac) or ipv6.is_dynamic(nm_ac)
```
We can return `True` directly, right? If ipv4 and ipv6 are disable this function is returning `False` which is not okay.

EdDev commented on libnmstate/nm/active_connection.py:
```diff
@@ -205,3 +197,27 @@ def _is_device_master_type(nmdev):
     return False
+
+
+def _ip_config_can_be_treat_as_activated(nm_dev, nm_ac):
```
This functions does too many things IMO and its name is not describing what it does, only how the caller interprets its output.

It will be nicer to read and change in the future if we can just read the full condition at the caller functions:
- If device reached ip config state AND
- this is a master device OR is_ip_dynamic(4) OR is_ip_dynamic(6) OR is_ip_dynamic(4,6)

is_ip_dynamic returns true only if both stacks have dynamic or if one has it and the other is READY with the other.

cathay4t replied:
Condition check has been moved back the the caller.

EdDev commented on libnmstate/nm/ipv6.py:
```diff
@@ -261,6 +261,16 @@ def is_dynamic(active_connection):
     return False
 
 
+def is_static(active_connection):
```
Does not seem to be used now.

EdDev commented on libnmstate/nm/ipv4.py:
```diff
@@ -182,6 +182,15 @@ def is_dynamic(active_connection):
     return False
 
 
+def is_static(active_connection):
```
Does not seem to be used now.

cathay4t commented:
New patch uploaded.

Changes:
 * Removed unused functions of `nm/ipv4.py` and `nm/ipv6.py`.
 * Reorganized the code to `self._ip_config_can_be_treat_as_activated()` and `self._reach_ip_config_state()`.
 * Added two more test case to test changing static IP to (dhcp with disabled ip stack).

ffmancera commented on libnmstate/nm/active_connection.py:
```diff
@@ -127,16 +127,22 @@ def is_active(self):
         if self.state == nm_acs.ACTIVATED:
             return True
         elif self.state == nm_acs.ACTIVATING:
+            ac_state_flags = nm_ac.get_state_flags()
```
I think this should be `nm_acs.get_state_flags()`, just a typo.

cathay4t replied:
Thanks.
